soroban-sdk/src/deploy.rs

```diff
@@ -54,85 +57,76 @@ impl Deployer {
-    /// from the current contract and the provided salt.
-    pub fn with_current_contract(
-        &self,
-        salt: &impl IntoVal<Env, BytesN<32>>,
-    ) -> DeployerWithCurrentContract {
-        let env = self.env();
-        DeployerWithCurrentContract {
-            env: env.clone(),
-            salt: salt.into_val(env),
+    /// from the current contract.
+    pub fn with_current_contract(&self) -> DeployerWithAddress {
+        DeployerWithAddress {
+            env: self.env.clone(),
+            address: self.env.current_contract_address(),
         }
     }
 
-    #[doc(hidden)]
-    /// Get a deployer for contracts that derive their contract IDs from the
-    /// given contract ID and the provided salt.
-    pub fn with_other_contract(
-        &self,
-        contract_id: &Address,
-        salt: &impl IntoVal<Env, BytesN<32>>,
-    ) -> DeployerWithOtherContract {
-        let env = self.env();
-        DeployerWithOtherContract {
-            env: env.clone(),
-            contract_id: contract_id.clone(),
-            salt: salt.into_val(env),
+    /// Get a deployer that deploys contracts that derive their contract IDs
+    /// from the provided address.
+    ///
+    /// The deployer address must authorize all the deployments.
+    pub fn with_address(&self, address: Address) -> DeployerWithAddress {
+        DeployerWithAddress {
+            env: self.env.clone(),
+            address,
         }
     }
+
+    /// Deploys a new instance of Stellar Asset Contract corresponding to
+    /// provided serialized asset.
+    ///
+    /// `serialized_asset` is the Stellar `Asset` XDR serialized to bytes. Refer
+    /// to `stellar-xdr` create for the exact definition.
+    pub fn deploy_stellar_asset(&self, serialized_asset: impl IntoVal<Env, Bytes>) -> Address {
+        self.env
+            .create_asset_contract(serialized_asset.into_val(&self.env).to_object())
+            .unwrap_infallible()
+            .into_val(&self.env)
+    }
+
+    /// Upload the contract Wasm code to the network.
+    ///
+    /// Returns the hash of the uploaded Wasm that can be then used for
+    /// the contract deployment.
+    pub fn upload_contract_wasm(&self, contract_wasm: impl IntoVal<Env, Bytes>) -> BytesN<32> {
+        self.env
+            .upload_wasm(contract_wasm.into_val(&self.env).to_object())
+            .unwrap_infallible()
+            .into_val(&self.env)
+    }
 }
 
-/// A deployer that deploys a contract that has its ID derived from the current
-/// contract ID and the provided salt.
-pub struct DeployerWithCurrentContract {
+/// A deployer that deploys a contract that has its ID derived from the provided
+/// address and salt.
+pub struct DeployerWithAddress {
     env: Env,
-    salt: BytesN<32>,
+    address: Address,
 }
 
-impl DeployerWithCurrentContract {
-    /// Return the ID of the contract defined by the deployer.
-    #[doc(hidden)]
-    pub fn id(&self) -> Address {
-        todo!()
-    }
```

Comment:
I don't think we need to remove this function, we're just about to fill this method in because @jayz22 just finished the environment changes to support it. The intent for why salt lived in the type represents a deployer which deploys to a specific address, not a variety of addresses. Could we keep salt there and forego the refactor?